File: mpc-back/src/models/Message.ts
// src/models/Message.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IMessage extends Document {
  client_id: mongoose.Types.ObjectId;
  content: string;
  fromShane: boolean;
  message_type: 'text' | 'image' | 'file' | 'audio';
  attachment?: {
    url: string;
    type: string;
    name?: string;
    size?: number;
  };
  timestamp: Date;
  status: {
    delivered?: Date;
    read?: Date;
  };
}

const MessageSchema = new Schema<IMessage>({
  client_id: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },
  content: {
    type: String,
    required: true
  },
  fromShane: {
    type: Boolean,
    required: true,
    index: true
  },
  message_type: {
    type: String,
    enum: ['text', 'image', 'file', 'audio'],
    default: 'text'
  },
  // ✅ CORRECT: attachment as nested object
  attachment: {
    url: { type: String },
    type: { type: String },
    name: { type: String },
    size: { type: Number },
  },
  timestamp: {
    type: Date,
    default: Date.now,
    index: true
  },
  status: {
    delivered: { type: Date },
    read: { type: Date },
  },
});

// Indexes
MessageSchema.index({ client_id: 1, timestamp: -1 });
MessageSchema.index({ client_id: 1, fromShane: 1, 'status.read': 1 });

export default mongoose.model<IMessage>('Message', MessageSchema);

File: mpc_admin_app/lib/app/models/message.dart
class Message {
  final String id;
  final String clientId;
  final String content;
  final bool fromShane;
  final DateTime timestamp;
  final String messageType;
  final Map<String, dynamic>? attachment;
  final MessageStatus status;

  Message({
    required this.id,
    required this.clientId,
    required this.content,
    required this.fromShane,
    required this.timestamp,
    this.messageType = 'text',
    this.attachment,
    required this.status,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['_id'],
      clientId: json['client_id'],
      content: json['content'] ?? '',
      fromShane: json['fromShane'],
      timestamp: DateTime.parse(json['timestamp']).toLocal(),
      messageType: json['message_type'] ?? 'text',
      attachment: json['attachment'],
      status: MessageStatus.fromJson(json['status'] ?? {}),
    );
  }
}

class MessageStatus {
  DateTime? delivered;
  DateTime? read;

  MessageStatus({this.delivered, this.read});

  factory MessageStatus.fromJson(Map<String, dynamic> json) {
    return MessageStatus(
      delivered:
          json['delivered'] != null ? DateTime.parse(json['delivered']) : null,
      read: json['read'] != null ? DateTime.parse(json['read']) : null,
    );
  }
}
